File: model_1/icpc-jakarta-2019/perf_1_solution.cpp
#include <iostream>
#include <vector>
#include <algorithm>
#include <climits>
using namespace std;
typedef long long ll;

class SegmentTree {
public:
    vector<ll> tree;
    vector<ll> lazy;
    int n;
    int size;

    SegmentTree(vector<ll>& data, int n_val) {
        n = n_val;
        size = 1;
        while (size < n) size *= 2;
        tree.assign(2 * size, (ll)1e18);
        lazy.assign(2 * size, 0);
        for (int i = 0; i < n; i++) {
            tree[size + i] = data[i];
        }
        for (int i = size - 1; i >= 1; i--) {
            tree[i] = min(tree[2 * i], tree[2 * i + 1]);
        }
    }

    void apply(int node, ll delta) {
        tree[node] += delta;
        if (node < size) {
            lazy[node] += delta;
        }
    }

    void push(int node) {
        if (lazy[node] != 0) {
            apply(2 * node, lazy[node]);
            apply(2 * node + 1, lazy[node]);
            lazy[node] = 0;
        }
    }

    void update_range(int l, int r, ll delta, int node, int segL, int segR) {
        if (r < segL || segR < l) return;
        if (l <= segL && segR <= r) {
            apply(node, delta);
            return;
        }
        push(node);
        int mid = (segL + segR) / 2;
        update_range(l, r, delta, 2 * node, segL, mid);
        update_range(l, r, delta, 2 * node + 1, mid + 1, segR);
        tree[node] = min(tree[2 * node], tree[2 * node + 1]);
    }

    void update(int l, int r, ll delta) {
        if (l > r) return;
        update_range(l, r, delta, 1, 0, size - 1);
    }

    ll global_min() {
        return tree[1];
    }
};

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0);

    int n, m, q;
    cin >> n >> m >> q;
    vector<ll> A(n);
    for (int i = 0; i < n; i++) {
        cin >> A[i];
    }

    ll A0 = A[0];
    int F1 = 0;
    for (int i = 0; i < n; i++) {
        if (A[i] < A0) {
            F1++;
        }
    }

    vector<int> R_arr(m + 1);
    vector<vector<ll>> B_arr(m + 1);
    vector<int> G_current(m + 1, 0);

    for (int i = 1; i <= m; i++) {
        cin >> R_arr[i];
        B_arr[i].resize(R_arr[i]);
        for (int j = 0; j < R_arr[i]; j++) {
            cin >> B_arr[i][j];
            if (B_arr[i][j] < A0) {
                G_current[i]++;
            }
        }
    }

    vector<ll> D_arr;
    if (m > 0) {
        D_arr.resize(m);
        ll S = F1;
        for (int i = 1; i <= m; i++) {
            D_arr[i - 1] = S - R_arr[i];
            S = S - R_arr[i] + G_current[i];
        }
    }

    SegmentTree* seg_tree = nullptr;
    if (m > 0) {
        seg_tree = new SegmentTree(D_arr, m);
    }

    vector<int> out_lines;
    for (int idx = 0; idx < q; idx++) {
        int X, Y;
        ll Z;
        cin >> X >> Y >> Z;
        Y--;

        ll old_val = B_arr[X][Y];
        int old_contrib = (old_val < A0) ? 1 : 0;
        int new_contrib = (Z < A0) ? 1 : 0;
        int delta = new_contrib - old_contrib;

        if (delta != 0) {
            G_current[X] += delta;
            if (X < m && seg_tree != nullptr) {
                seg_tree->update(X, m - 1, delta);
            }
            B_arr[X][Y] = Z;
        }

        if (m == 0) {
            out_lines.push_back(1);
        } else {
            ll min_val = seg_tree->global_min();
            if (min_val >= 0) {
                out_lines.push_back(1);
            } else {
                out_lines.push_back(0);
            }
        }
    }

    for (int i = 0; i < out_lines.size(); i++) {
        cout << out_lines[i] << '\n';
    }

    if (seg_tree != nullptr) {
        delete seg_tree;
    }

    return 0;
}

File: model_1/icpc-jakarta-2019/perf_1_solution.py
import sys

class SegmentTree:
    def __init__(self, data, n):
        self.n = n
        self.size = 1
        while self.size < n:
            self.size *= 2
        self.tree = [10**18] * (2 * self.size)
        self.lazy = [0] * (2 * self.size)
        for i in range(n):
            self.tree[self.size + i] = data[i]
        for i in range(n, self.size):
            self.tree[self.size + i] = 10**18
        for i in range(self.size - 1, 0, -1):
            self.tree[i] = min(self.tree[2*i], self.tree[2*i+1])
    
    def apply(self, node, delta):
        self.tree[node] += delta
        if node < self.size:
            self.lazy[node] += delta
            
    def push(self, node):
        if self.lazy[node] != 0:
            self.apply(2*node, self.lazy[node])
            self.apply(2*node+1, self.lazy[node])
            self.lazy[node] = 0
            
    def update_range(self, l, r, delta, node, segL, segR):
        if r < segL or segR < l:
            return
        if l <= segL and segR <= r:
            self.apply(node, delta)
            return
        self.push(node)
        mid = (segL + segR) // 2
        self.update_range(l, r, delta, 2*node, segL, mid)
        self.update_range(l, r, delta, 2*node+1, mid+1, segR)
        self.tree[node] = min(self.tree[2*node], self.tree[2*node+1])
        
    def update(self, l, r, delta):
        if l > r:
            return
        self.update_range(l, r, delta, 1, 0, self.size-1)
        
    def global_min(self):
        return self.tree[1]

def main():
    data = sys.stdin.read().split()
    if not data:
        return
    
    it = iter(data)
    n = int(next(it)); m = int(next(it)); q = int(next(it))
    A = [int(next(it)) for _ in range(n)]
    A0 = A[0]
    F1 = sum(1 for x in A if x < A0)
    
    R_arr = [0] * (m+1)
    B_arr = [None] * (m+1)
    G_current = [0] * (m+1)
    
    for i in range(1, m+1):
        R_i = int(next(it))
        R_arr[i] = R_i
        B_i = [int(next(it)) for _ in range(R_i)]
        B_arr[i] = B_i
        count_less = sum(1 for val in B_i if val < A0)
        G_current[i] = count_less
        
    if m > 0:
        H = [0] * (m+2)
        H[1] = F1
        D0 = [0] * m
        for t in range(1, m+1):
            D0[t-1] = H[t] - R_arr[t]
            H[t+1] = H[t] - R_arr[t] + G_current[t]
        seg_tree = SegmentTree(D0, m)
    else:
        seg_tree = None
        
    out_lines = []
    for _ in range(q):
        X = int(next(it)); Y = int(next(it)); Z = int(next(it))
        idx = Y - 1
        old_val = B_arr[X][idx]
        old_contrib = 1 if old_val < A0 else 0
        new_contrib = 1 if Z < A0 else 0
        delta = new_contrib - old_contrib
        
        if delta != 0:
            if m > 0 and X <= m-1:
                seg_tree.update(X, m-1, delta)
            B_arr[X][idx] = Z
            G_current[X] += delta
            
        if m == 0:
            out_lines.append("1")
        else:
            min_val = seg_tree.global_min()
            if min_val >= 0:
                out_lines.append("1")
            else:
                out_lines.append("0")
                
    sys.stdout.write("\n".join(out_lines) + "\n")

if __name__ == "__main__":
    main()